src/std/misc/text.ss

```diff
@@ -13,3 +19,147 @@
      (let (path (gx#core-resolve-path #'path (stx-source stx)))
        (with-syntax ((txt (call-with-input-file path read-all-as-string)))
          #'(quote txt))))))
+
+;; quasistring utility functions
+(begin-syntax
+  (def qs-err-missing-close "Malformed quasistring: closing } required after s-exp")
+
+  (def (stx-format stx)
+    (if (stx-string? stx) ;; don't format strings
+      stx
+      (with-syntax ((stx stx))
+        #'(format "~a" stx))))
+
+  (def (map-format stx)
+    (stx-map stx-format stx))
+
+  ;; custom eval to capture the proper variables
+  (def (eval** stx ctx)
+    (parameterize ((current-expander-phi (1+ (current-expander-phi))))
+      (let (stx (datum->syntax ctx stx))
+        (eval-syntax stx #t))))
+
+  (def (parse-quasistring stx ctx rt-symbol-proc output-proc (expansion-time-eval #f))
+    ;; rt-symbol-proc wraps sexp if sexp is a run time symbol
+    ;; output-proc is what we return from the macro
+    ;; if expansion-time-eval is #t, allow for ##{} template variables
+    (def (lp (port (open-input-string (stx-e stx)))
```

Comment:
Here we open the string as a port to go through it one char at a time using `read-char`.